File: builder/Builder.py
from typing import Protocol


class Laptop():
    cpu: str
    ram: str
    os: str

    def __init__(self):
        self.cpu = ""
        self.ram = ""
        self.os = ""

    def setCpu(self, cpu: str):
        self.cpu = cpu

    def setRam(self, ram: str):
        self.ram = ram

    def setOs(self, os: str):
        self.os = os

    def describe(self):
        print("CPU: " + self.cpu)
        print("RAM: " + self.ram)
        print("OS: " + self.os)


class LaptopBuilder(Protocol):
    laptop: Laptop

    def setCpu(self):
        pass

    def setRam(self):
        pass

    def setOs(self):
        pass

    def build(self) -> Laptop:
        return self.laptop


class HighPerformanceLaptopBuilder(LaptopBuilder):
    laptop: Laptop

    def __init__(self):
        self.laptop = Laptop()

    def setCpu(self):
        self.laptop.setCpu("i7 CPU")

    def setRam(self):
        self.laptop.setRam("16GB RAM")

    def setOs(self):
        self.laptop.setOs("Windows 10")

    def build(self) -> Laptop:
        return self.laptop


class LowPerformanceLaptopBuilder(LaptopBuilder):
    laptop: Laptop

    def __init__(self):
        self.laptop = Laptop()

    def setCpu(self):
        self.laptop.setCpu("i3 CPU")

    def setRam(self):
        self.laptop.setRam("4GB RAM")

    def setOs(self):
        self.laptop.setOs("Windows 7")

    def build(self) -> Laptop:
        return self.laptop


class Director:
    laptopBuilder: LaptopBuilder

    def setLaptopBuilder(self, laptopBuilder: LaptopBuilder):
        self.laptopBuilder = laptopBuilder

    def getLaptop(self):
        self.laptopBuilder.setCpu()
        self.laptopBuilder.setRam()
        self.laptopBuilder.setOs()
        return self.laptopBuilder.build()


def testLaptopBuilder():
    director = Director()
    highPerformanceLaptopBuilder = HighPerformanceLaptopBuilder()
    director.setLaptopBuilder(highPerformanceLaptopBuilder)
    highPerformanceLaptop: Laptop = director.getLaptop()
    print("High performance laptop:")
    highPerformanceLaptop.describe()

    lowPerformanceLaptopBuilder = LowPerformanceLaptopBuilder()
    director.setLaptopBuilder(lowPerformanceLaptopBuilder)
    lowPerformanceLaptop = director.getLaptop()
    print("Low performance laptop:")
    lowPerformanceLaptop.describe()


if __name__ == "__main__":
    testLaptopBuilder()

File: builder/Builder.cpp
#include <iostream>
#include <string>

using namespace std;

class Laptop {
    private:
        string cpu;
        string ram;
        string os;
    public:
        void setCpu(string cpu) {
            this->cpu = cpu;
        }
        void setRam(string ram) {
            this->ram = ram;
        }
        void setOs(string os) {
            this->os = os;
        }
        void describe() {
            cout << "CPU: " << cpu << endl;
            cout << "RAM: " << ram << endl;
            cout << "OS: " << os << endl;
        }
};

class LaptopBuilder {
    public:
        virtual void setCpu() = 0;
        virtual void setRam() = 0;
        virtual void setOs() = 0;
        virtual Laptop* build() = 0;
};

class HighPerformanceLaptopBuilder : public LaptopBuilder {
    private:
        Laptop* highPerformanceLaptop;
    public:
        HighPerformanceLaptopBuilder() {
            highPerformanceLaptop = new Laptop();
        }
        void setCpu() {
            highPerformanceLaptop->setCpu("i7 CPU");
        }
        void setRam() {
            highPerformanceLaptop->setRam("16GB RAM");
        }
        void setOs() {
            highPerformanceLaptop->setOs("Windows 10");
        }
        Laptop* build() {
            return highPerformanceLaptop;
        }
};

class LowPerformanceLaptopBuilder : public LaptopBuilder {
    private:
        Laptop* lowPerformanceLaptop;
    public:
        LowPerformanceLaptopBuilder() {
            lowPerformanceLaptop = new Laptop();
        }
        void setCpu() {
            lowPerformanceLaptop->setCpu("i3 CPU");
        }
        void setRam() {
            lowPerformanceLaptop->setRam("4GB RAM");
        }
        void setOs() {
            lowPerformanceLaptop->setOs("Windows 7");
        }
        Laptop* build() {
            return lowPerformanceLaptop;
        }
};

class Director {
    private:
        LaptopBuilder* laptopBuilder;
    public:
        void setLaptopBuilder(LaptopBuilder* laptopBuilder) {
            this->laptopBuilder = laptopBuilder;
        }
        Laptop* getLaptop() {
            laptopBuilder->setCpu();
            laptopBuilder->setRam();
            laptopBuilder->setOs();
            return laptopBuilder->build();
        }
};

int main() {
    Director* director = new Director();
    LaptopBuilder* highPerformanceLaptopBuilder = new HighPerformanceLaptopBuilder();
    director->setLaptopBuilder(highPerformanceLaptopBuilder);
    Laptop* highPerformanceLaptop = director->getLaptop();
    cout << "High performance laptop:" << endl;
    highPerformanceLaptop->describe();

    LaptopBuilder* lowPerformanceLaptopBuilder = new LowPerformanceLaptopBuilder();
    director->setLaptopBuilder(lowPerformanceLaptopBuilder);
    Laptop* lowPerformanceLaptop = director->getLaptop();
    cout << "Low performance laptop:" << endl;
    lowPerformanceLaptop->describe();
}

